### PR TITLE
feat(agw): Add support for IPv6 Non-NAT router mode.

### PR DESCRIPTION
**Pull request by pshelar**
This add support to set IPv6 add and IPv6 GW IP for SGi
device on AGW.
It also cleans up test cases for uplink-br app. As a result
two skipped test can be enabled.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

## Test Plan
validated using new and existing test cases on magma dev vm.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py [hunk 1]
```diff
@@ -619,10 +630,12 @@ def setUpClass(cls):
             cls.UPLINK_BRIDGE,
             cls.UPLINK_PATCH, None,
         )
-
+        print("check ipv6")
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py]: Rm these prints

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 2]
```diff
@@ -230,6 +238,18 @@ def _set_sgi_interface_ingress_flows(self):
             actions = "output:LOCAL"
             self._install_flow(flows.MEDIUM_PRIORITY + 1, match, actions)
 
+        if_addrs = netifaces.ifaddresses(self.config.uplink_bridge).get(netifaces.AF_INET6, [])
+        for addr in if_addrs:
+            addr = addr['addr']
+            if '%' in addr:
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]: if % not in addr what happens here? Lets either check the return(if not % in addr: exit) or remove this if ?

File: lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py [hunk 3]
```diff
@@ -801,22 +825,145 @@ def test_flow_snapshot_match(self):
         subprocess.check_call(cmd1)
         assert_bridge_snapshot_match(self, self.BRIDGE, self.service_manager)
 
+#----
+# DHCP based test
+@unittest.skip
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py:822]: why are we skipping this one?

**Review by koolzz — APPROVED** [review]

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py (hunk 2 above)
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/uplink_bridge.py]: its optional, so it does not matter.

File: lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py (hunk 3 above)
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py:822]: this depends on a thread running in uplink-br app. so the test is stuck in hung state. I need to figure out how to run the test in parallel.

Comment (pshelar) [reply lte/gateway/python/magma/pipelined/tests/test_uplink_bridge.py:822]: I will do it as followup.

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 4]
```diff
@@ -230,6 +238,18 @@ def _set_sgi_interface_ingress_flows(self):
             actions = "output:LOCAL"
             self._install_flow(flows.MEDIUM_PRIORITY + 1, match, actions)
 
+        if_addrs = netifaces.ifaddresses(self.config.uplink_bridge).get(netifaces.AF_INET6, [])
+        for addr in if_addrs:
+            addr = addr['addr']
+            # get rid of dev name if present.
+            addr = addr.split('%')[0]
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py:245]: should we add `if % in addr` check?

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 5]
```diff
@@ -389,21 +429,33 @@ def _set_sgi_ip_addr(self, if_name: str):
...
+    def _flush_ip(self, if_name, af_inet: int):
+        ver = "-4"
+        if af_inet == netifaces.AF_INET6:
+            ver = "-6"
+        flush_ip = ["ip", ver, "addr", "flush", "dev", if_name]
         subprocess.check_call(flush_ip)
 
-    def _kill_dhclient(self, if_name):
+    def _cleanup_ipv4_and_ipv6(self, if_name: str):
+        self._cleanup_if(if_name, True, netifaces.AF_INET)
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]: add keyword args for these
```suggestion
        self._cleanup_if(if_name=if_name, flush=True, af_inet=netifaces.AF_INET)
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 6]
```diff
@@ -389,21 +429,33 @@ def _set_sgi_ip_addr(self, if_name: str):
...
+        self._cleanup_if(if_name, True, netifaces.AF_INET6)
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]: same here, add keyword args

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 7]
```diff
@@ -389,21 +429,33 @@ def _set_sgi_ip_addr(self, if_name: str):
...
+
+    def _kill_dhclient(self, if_name, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _kill_dhclient(self, if_name: str, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 8]
```diff
@@ -389,21 +429,33 @@ def _set_sgi_ip_addr(self, if_name: str):
         except subprocess.SubprocessError as e:
             self.logger.warning("Error while setting SGi IP: %s", e)
 
-    def _is_iface_ip_set(self, if_name, ip_addr):
+    def _is_iface_ip_set(self, if_name, ip_addr, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _is_iface_ip_set(self, if_name, ip_addr, af_inet: int) -> bool:
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 9]
```diff
@@ -412,9 +464,12 @@ def _kill_dhclient(self, if_name):
         for pid in pgrep_out.stdout.readlines():
             subprocess.check_call(["kill", pid.strip()])
 
-    def _restart_dhclient(self, if_name):
+    def _restart_dhclient(self, if_name, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _restart_dhclient(self, if_name: str, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 10]
```diff
@@ -435,8 +490,10 @@ def _setup_vlan_pop_dev(self):
                 self.config.dev_vlan_out, "71",
             )
 
-    def _cleanup_if(self, if_name, flush: bool):
-        # Release eth IP first.
+    def _release_dhcp_ip(self, if_name, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _release_dhcp_ip(self, if_name: str, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 11]
```diff
@@ -446,17 +503,23 @@ def _cleanup_if(self, if_name, flush: bool):
                 release_eth_ip, ex,
             )
 
+    def _cleanup_if(self, if_name, flush: bool, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _cleanup_if(self, if_name: str, flush: bool, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 12]
```diff
@@ -446,17 +503,23 @@ def _cleanup_if(self, if_name, flush: bool):
...
-            self._flush_ip(if_name)
+            self._flush_ip(if_name, af_inet)
         except subprocess.CalledProcessError as ex:
             self.logger.info("could not flush ip addr: %s, %s", if_name, ex)
 
         self.logger.info("SGi DHCP: port [%s] ip removed", if_name)
 
-    def _restart_dhclient_if(self, if_name):
-        self._cleanup_if(if_name, False)
+    def _start_dhclient_if(self, if_name, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _start_dhclient_if(self, if_name: str, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 13]
```diff
@@ -467,10 +530,16 @@ def _restart_dhclient_if(self, if_name):
                 setup_dhclient, ex,
             )
 
+    def _restart_dhclient_if(self, if_name, af_inet: int):
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]:
```suggestion
    def _restart_dhclient_if(self, if_name: str, af_inet: int):
```

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py [hunk 14]
```diff
@@ -467,10 +530,16 @@ def _restart_dhclient_if(self, if_name):
...
+        self._cleanup_if(if_name, False, af_inet)
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/uplink_bridge.py]: add keyword args

**Review by ardzoht — CHANGES_REQUESTED** [review]
some nits on formatting and str split check

File: lte/gateway/python/magma/pipelined/app/uplink_bridge.py (hunk 4 above)
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/uplink_bridge.py:245]: no need, since we are looking at first element of array any ways.

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**